examples/Cassie/cassie_rbt_state_estimator.cc

```diff
@@ -816,32 +848,164 @@ void CassieRbtStateEstimator::contactEstimation(
+    int* left_contact, int* right_contact) const {
+  // If all three costs are high, we believe it's going through impact event (
+  // big ground contact point acceleration),
+  // and we assume there is no support legs because we don't want moving feet
+  // to mess up EKF.
+  if ((optimal_cost->at(0) >= cost_threshold_ekf_) &&
+      (optimal_cost->at(1) >= cost_threshold_ekf_) &&
+      (optimal_cost->at(2) >= cost_threshold_ekf_)) {
+    *left_contact = 0;
+    *right_contact = 0;
+  } else {
+    if (print_info_to_terminal_) {
+      cout << "optimal_cost[0][1][2], threshold = " <<
+          optimal_cost->at(0) << ", " << optimal_cost->at(1) << ", " <<
+          optimal_cost->at(2) << ", " << cost_threshold_ekf_ << endl;
+      cout << "left/right contacts = " <<
+          *left_contact << ", " << *right_contact << endl;
+    }
+
+    // Only use spring to determine the contact (we want to say a toe is on the
+    // ground when it is *flat* on the ground.)
+    // We say a foot is in contact with the ground if knee and heel spring
+    // deflections are *both* over some thresholds. We don't update anything
+    // if it's under the threshold.
+    double left_knee_spring = output.GetPositionAtIndex(
+        position_index_map_.at("knee_joint_left"));
+    double right_knee_spring = output.GetPositionAtIndex(
+        position_index_map_.at("knee_joint_right"));
+    double left_heel_spring = output.GetPositionAtIndex(
+        position_index_map_.at("ankle_spring_joint_left"));
+    double right_heel_spring = output.GetPositionAtIndex(
+        position_index_map_.at("ankle_spring_joint_right"));
+    if (left_knee_spring < knee_spring_threshold_ekf_ &&
+          left_heel_spring < heel_spring_threshold_ekf_) {
+      *left_contact = 1;
+    }
+    if (right_knee_spring < knee_spring_threshold_ekf_ &&
+          right_heel_spring < heel_spring_threshold_ekf_) {
+      *right_contact = 1;
+    }
+
+    if (print_info_to_terminal_) {
+      cout << "left/right knee spring, threshold = " <<
+          left_knee_spring << ", " << right_knee_spring << ", " <<
+          knee_spring_threshold_ekf_ << endl;
+      cout << "left/right heel spring, threshold = " <<
+          left_heel_spring << ", " << right_heel_spring << ", " <<
+          heel_spring_threshold_ekf_ << endl;
+      cout << "left/right contacts = " <<
+          *left_contact << ", " << *right_contact << endl;
+    }
+  }
+}
+
+
+/// EstimateContactForController(). Less conservative.
+///
+/// Input:
+///  - OutputVector `output` containing the positions, velocities and
+///    actuator torques of the robot
+/// Ouput: left contact `left_contact` and right contact `right_contact` that
+///  indicate if the corresponding foot is in contact with the ground
+///
+/// Assumptions:
+///  1. the swing leg doesn't stop during single support
+///  2. flight mode is not present during the gait
+///
+/// Algorithm:
+///  The contact is estimated based on
```

Comment:
Same as above comment regarding the algorithm

Reply:
I've updated the logic so that it's using the result from double support QP